source/common/upstream/subset_lb.cc

```diff
@@ -62,14 +62,19 @@ SubsetLoadBalancer::SubsetLoadBalancer(
     panic_mode_subset_ = subset_any_;
   }
 
+  initSubsetSelectorMap();
+
   // Create filtered default subset (if necessary) and other subsets based on current hosts.
   refreshSubsets();
 
-  initSubsetSelectorMap();
+  // This must happen after `initSubsetSelectorMap()` because that initializes `single_`.
+  rebuildSingle();
 
   // Configure future updates.
   original_priority_set_callback_handle_ = priority_set.addPriorityUpdateCb(
       [this](uint32_t priority, const HostVector& hosts_added, const HostVector& hosts_removed) {
+        rebuildSingle();
```

Comment:
I think this is O(n^2) (worst case being one host per priority, then each host is updated). Maybe add a comment/TODO here in case there ever are performance issues here?

Reply:
Oh, good catch! I only tested with a single priority.